src/fsharp/FSharp.Core/async.fs

```diff
@@ -1250,31 +1250,28 @@ namespace Microsoft.FSharp.Control
                                 |> unfake)
                     | Some maxDegreeOfParallelism ->
                         let mutable i = -1
-                        let worker = MakeAsync (fun _ ->
-                            while i < tasks.Length do
+                        let rec worker (trampolineHolder : TrampolineHolder) =
+                            if i < tasks.Length then
                                 let j = Interlocked.Increment &i
                                 if j < tasks.Length then
-                                    let trampolineHolder = new TrampolineHolder()
-                                    trampolineHolder.ExecuteWithTrampoline (fun () ->
-                                        let ctxt =
+                                    if innerCTS.Token.IsCancellationRequested then
+                                        let cexn = new OperationCanceledException (innerCTS.Token)
+                                        recordFailure (Choice2Of2 cexn) |> unfake
```

Comment:
Why do you need to trigger the continuation continuation here? innerCts is triggered either by the exception continuation (so it has already been run) or the parent token has been cancelled.

Reply:
@eiriktsarpalis,

Yes … if j is not less than tasks.Length then there is no work to do. So no interlock is necessary.  If a thread switch happened after the first if and before the interlocked increment j would have a different value from i and so the additional check is necessary to see if work needs to be done.

Reply:
If the token is cancelled after we've entered the ProtectedCode section but before any workers run then we need this to trigger cancellation.